### PR TITLE
Fix `require-default-props` rule when using Flow type from assignment

### PR DESCRIPTION
**Pull request by wyze**
This fixes the issue I reported in #1017.

The condition I added is hit when code like the following is used:

```js
type Props = Message
```

This just takes the right side of the annotation (`Message`) and tries to find that variable in scope and then continues on.

### CODE REVIEW

**Comment by wyze** [comment]
Hey @EvNaverniouk, can you check the failing case I added? Seems a bit off as I needed to provide `errors: []` because it is under `invalid` key in the test suite, but it doesn't actually output any errors.

File: tests/lib/rules/require-default-props.js [hunk 1]
```diff
@@ -1718,6 +1728,17 @@ ruleTester.run('require-default-props', rule, {
...
+    {
+      code: [
+        'import type { ImportedType } from "fake";',
+        'type Props = ImportedType;',
+        'function Hello(props: Props) {',
+        '  return <div>Hello {props.name.firstname}</div>;',
+        '}'
+      ].join('\n'),
+      parser: 'babel-eslint',
+      errors: []
```
Comment (EvNaverniouk) [inline tests/lib/rules/require-default-props.js]: We should update this to return errors so we can test for it, perhaps in the event that the type isn't defined. Maybe something like this:

```
function Hello (props: UndefinedVariable) {
    return <div>Hello {props.name.firstname}</div>;
}
```

Comment (wyze) [inline tests/lib/rules/require-default-props.js]: When you say "update this", are you referring to the rule, to throw an error if it can't find a variable? If so, I think that should be a separate PR.

If you are referring to the test case when you say "update this", then when making the requested changes, the invalid test still passes with `errors: []` so I am not sure there is much difference between what you are suggesting and what I added.

Comment (EvNaverniouk) [inline tests/lib/rules/require-default-props.js]: Perhaps I'm misunderstanding what the scope of the changes in this PR. I was under the impression that it adds support for Flow types as valid prop assignments via variables. Your original test case covers the fact the the rule properly catches such valid Flow type assignments without error.

What I'm asking for is to add a test case that when a Flow type is defined, but the Flow type is variable is unassigned. I believe this ESLint rule should still throw an error in that case.

Comment (ljharb) [inline tests/lib/rules/require-default-props.js]: That's covered by the no-undef rules, is it not?

Comment (yannickcr) [inline tests/lib/rules/require-default-props.js]: Yes, this should be covered by no-undef. The require-default-props rule should just try its bests to resolve the props and the defaultProps, and give up if it cannot do it.

If this test case do not return any error then it should be in the "valid" array.

File: lib/rules/require-default-props.js [hunk 2]
```diff
@@ -141,6 +141,11 @@ module.exports = {
       switch (node.typeAnnotation.type) {
         case 'GenericTypeAnnotation':
           var annotation = resolveGenericTypeAnnotation(node.typeAnnotation);
+
+          if (annotation && annotation.id) {
+            annotation = findVariableByName(annotation.id.name);
+          }
+
           properties = annotation ? annotation.properties : [];
```
Comment (yannickcr) [inline lib/rules/require-default-props.js]: You should also update the next line to check for `annotation.properties` existence, or the rule will keep crashing if there is multiple reassignments

```js
properties = annotation && annotation.properties ? annotation.properties : [];
```

Comment (ljharb) [inline lib/rules/require-default-props.js]: What would be a test case that would trigger that?

Comment (wyze) [inline lib/rules/require-default-props.js]: Well, technically, my fix prevented this, but if for some reason you do multiple assignments, it would trigger the original error again:

```js
import type ImportedProps from "fake";
type NestedProps = ImportedProps;
type Props = NestedProps;
function Hello(props: Props) {
  return <div>Hello {props.name.firstname}</div>;
}
```

**Comment by cr0cK** [comment]
up!
